File: LAB04/LAB04.c
#define _USE_MATH_DEFINES
#include <stdio.h>
#include <math.h>

int main() {

int x[]={0, 0, 0, 1, 1}; //3 вариант
int y[]={0, 1, 0, 1, 0}; //3+7=10
int randmass[31]={0}; //послед Голда

	for (int i=0; i<31; i++) {

	    int xlast=0;
	    int ylast=0;

	    if (x[0] != x[2]) {
		xlast=1;
		}
	    if (y[1] != y[3]) {
		ylast=1;
		}

	    for (int j=5; j>0; j--) {

		x[j]=x[j-1];
		y[j]=y[j-1];

		}
	    x[0]=xlast;
	    y[0]=ylast;

	    int vivod=0;

	    if (x[4] != y[4]) {

		vivod=1;

		}
	    randmass[i]=vivod; //массив вывода
	    printf("%2d : x[]={%d %d %d %d %d} y[]={%d %d %d %d %d} vivod=%d randmass=%d\n", i, x[0], x[1], x[2], x[3], x[4], y[0], y[1], y[2], y[3], y[4], vivod, randmass[i]);

	}

int x1[]={0, 0, 1, 0, 0}; //3+1=4
int y1[]={0, 0, 1, 0, 1}; //10-5=5
int randmass1[31]={0};

for (int i=0; i<31; i++) { //2 послед Голда

            int xlast=0;
            int ylast=0;

            if (x1[0] != x1[2]) {
                xlast=1;
                }
            if (y1[1] != y1[3]) {
                ylast=1;
                }

            for (int j=5; j>0; j--) {

                x1[j]=x1[j-1];
                y1[j]=y1[j-1];

                }
            x1[0]=xlast;
            y1[0]=ylast;

            int vivod=0;

            if (x1[4] != y1[4]) {

                vivod=1;

                }
	    randmass1[i]=vivod; //вывод
            printf("%2d : x1[]={%d %d %d %d %d} y1[]={%d %d %d %d %d} vivod=%d randmass1=%d\n", i, x1[0], x1[1], x1[2], x1[3], x1[4], y1[0], y1[1], y1[2], y1[3], y1[4], vivod, randmass1[i]);

        }
printf("                                                                 Corr x, y\n");
printf(" С| B1| B2| B3| B4| B5| B6| B7| B8| B9|B10|B11|B12|B13|B14|B15|B16|B17|B18|B19|B20|B21|B22|B23|B24|B25|B26|B27|B28|B29|B30|B31|А.Kор\n");

int vivodlast[31]={0};
        for (int m=0; m<31; m++) {
                vivodlast[m]=randmass[m];
        }
for (int k=0; k<36; k++) {
	printf("%2d", k);
	if (k != 0) {
	int lastvivod=randmass[30];
        for (int n=1; n<31; n++) {
		randmass[31-n]=randmass[31-n-1];
        }
	randmass[0]=lastvivod;
	}
	float sumcorr=0.0;
	float sumuncorr=0.0;
	for (int l=0; l<31; l++) {
                printf("|%3d", randmass[l]);
		if (randmass[l]==vivodlast[l]) {
			sumcorr += 1.0;
		}
		else {
			sumuncorr += 1.0;
		}
        }
	float autocorr=(1/31.0)*(sumcorr-sumuncorr);
	printf("|%.2f\n", autocorr);
}
//Корреляция между последов Голда
float rhoxy = 0.00;

for (int i=0; i<=30; i++) {
      rhoxy += randmass[i]*randmass1[i];
}
printf("rhoxy=%.f\n", rhoxy);
return 0;
}

File: LAB04/lab04.m
x = [0, 0, 0, 1, 1]; %3
y = [0, 1, 0, 1, 0]; %10
randmass = zeros(1, 31);
for i = 1:31 %послед Голда
    xlast = 0;
    ylast = 0;
    if x(1) ~= x(3)
        xlast = 1;
    end
    if y(2) ~= y(4)
        ylast = 1;
    end
    for j = 5:-1:2
        x(j) = x(j-1);
        y(j) = y(j-1);
    end
    x(1) = xlast;
    y(1) = ylast;
    vivod = 0;
    if x(5) ~= y(5)
        vivod = 1;
    end
    randmass(i) = vivod;
    fprintf('%2d : x=[%d %d %d %d %d] y=[%d %d %d %d %d] vivod=%d randmass=%d\n', i, x(1), x(2), x(3), x(4), x(5), y(1), y(2), y(3), y(4), y(5), vivod, randmass(i));
end

x1 = [0, 0, 1, 0, 0]; %4
y1 = [0, 0, 1, 0, 1]; %5
randmass1 = zeros(1, 31);
for i = 1:31 %Голд
    xlast = 0;
    ylast = 0;
    if x1(1) ~= x1(3)
        xlast = 1;
    end
    if y1(2) ~= y1(4)
        ylast = 1;
    end
    for j = 5:-1:2
        x1(j) = x1(j-1);
        y1(j) = y1(j-1);
    end
    x1(1) = xlast;
    y1(1) = ylast;
    vivod = 0;
    if x1(5) ~= y1(5)
        vivod = 1;
    end
    randmass1(i) = vivod;
    fprintf('%2d : x1=[%d %d %d %d %d] y1=[%d %d %d %d %d] vivod=%d randmass1=%d\n', i, x1(1), x1(2), x1(3), x1(4), x1(5), y1(1), y1(2), y1(3), y1(4), y1(5), vivod, randmass1(i));
end
figure
subplot(2, 1, 1);
fprintf('                                                             Corr x, y in Matlab with xcorr\n');
fprintf(' С| B1| B2| B3| B4| B5| B6| B7| B8| B9|B10|B11|B12|B13|B14|B15|B16|B17|B18|B19|B20|B21|B22|B23|B24|B25|B26|B27|B28|B29|B30|B31|А.Kор\n');
vivodlast = randmass;
autocorr_values = zeros(1, 36);
shifts = 1:36;
for k = 1:36
    fprintf('%2d', k);
    if k ~= 1
        lastvivod = randmass(31);
        for n = 1:30
            randmass(32-n) = randmass(31-n);
        end
        randmass(1) = lastvivod;
    end
    [c, lags] = xcorr(randmass, vivodlast, 35, 'normalized');
    autocorr_values(k) = c(36);
    fprintf('|%3d', randmass);
    fprintf('|%.2f\n', autocorr_values(k));
end

    plot(shifts-1, autocorr_values, '-o');
    title("Autocorr in MatLAB");
    grid on;
%%%%%%%%%%%%%%%%%%%%%%%%%%%
subplot(2, 1, 2);
hold on;
fprintf('                                                                   Corr x, y in C\n');
fprintf(' С| B1| B2| B3| B4| B5| B6| B7| B8| B9|B10|B11|B12|B13|B14|B15|B16|B17|B18|B19|B20|B21|B22|B23|B24|B25|B26|B27|B28|B29|B30|B31|А.Kор\n');
randmass = vivodlast;
autocorr_values=zeros(36, 1);

vivodlast = randmass;
shifts = 1:36;
for k = 1:36
    fprintf('%2d', k);
    if k ~= 1
        lastvivod = randmass(31);
        for n = 1:30
            randmass(32-n) = randmass(31-n);
        end
        randmass(1) = lastvivod;
    end
    sumcorr = 0;
    sumuncorr = 0;
    for l = 1:31
        fprintf('|%3d', randmass(l));
        if randmass(l) == vivodlast(l)
            sumcorr = sumcorr + 1;
        else
            sumuncorr = sumuncorr + 1;
        end
    end
    autocorr = (1/31)*(sumcorr - sumuncorr);
    fprintf('|%.2f\n', autocorr);
    autocorr_values(k) = autocorr;
end
plot(shifts-1, autocorr_values, 'o-');
title("Autocorr in C");
grid on;
%%%%%%%%%%%%%%%%%%%%%%%%%%% коррел между двумя послед Голда
r = xcorr(randmass, randmass1);
fprintf('rhoxy = ');
disp(r(32));